File: kalodata_loader_service/tempmail.py
import requests


class Email:
    def __init__(self, login: str, domain: str, eid: int, sender: str, do_action):
        self.__login = login
        self.__domain = domain
        self.__eid = eid
        self.__sender = sender
        self.do_action = do_action
        self.__data = None

    def fetch_data(self) -> dict:
        try:
            data = self.do_action({"action": "readMessage", "login": self.__login, "domain": self.__domain, "id": self.__eid})

            self.__data = {
                'id': self.__eid,
                'sender': self.__sender,
                'body': data['body'],
                'text': data['textBody'],
                'html': data['htmlBody'],
            }

            return self.__data
        except Exception as e:
            raise Exception(f"Error have occurred while fetching Email[eid={self.__eid}] data") from e


class TempMail:
    def __init__(self):
        self.req_client = requests.session()

        self.email: str = ''
        self.login: str = ''
        self.domain: str = ''

        self.gen_email()

    def gen_email(self) -> None:
        try:
            self.email = self.do_action({"action": 'genRandomMailbox'})[0]
            self.login = self.email.split('@')[0]
            self.domain = self.email.split('@')[1]
        except Exception as e:
            raise Exception('Error have occurred while generating email') from e

    def do_action(self, params: dict):
        return self.req_client.get('https://www.1secmail.com/api/v1/', params=params).json()

    def fetch_emails(self, _from: str | None = None) -> list[Email]:
        try:
            dat = self.do_action({"action": "getMessages", "login": self.login, "domain": self.domain})
            if _from:
                return [Email(self.login, self.domain, ml['id'], ml['from'], self.do_action)
                        for ml in dat if _from in ml['from']]
            return [Email(self.login, self.domain, ml['id'], ml['from'], self.do_action) for ml in dat]
        except Exception as e:
            raise Exception('Error have occurred while fetching emails') from e

File: kalodata_loader_service/tempmail.js
class Email {
    constructor(login, domain, eid, sender, doAction) {
        this._login = login;
        this._domain = domain;
        this._eid = eid;
        this._sender = sender;
        this.doAction = doAction;
        this._data = null;
    }

    async fetchData() {
        try {
            const data = await this.doAction({
                "action": "readMessage",
                "login": this._login,
                "domain": this._domain,
                "id": this._eid
            });

            this._data = {
                'id': this._eid,
                'sender': this._sender,
                'body': data.body,
                'text': data.textBody,
                'html': data.htmlBody,
            };

            return this._data;
        } catch (e) {
            throw new Error(`Error have occurred while fetching Email[eid=${this._eid}] data: ${e.message}`);
        }
    }
}

class TempMail {
    constructor() {
        this.email = '';
        this.login = '';
        this.domain = '';
    }

    async genEmail() {
        try {
            const response = await fetch('https://www.1secmail.com/api/v1/?action=genRandomMailbox');
            const data = await response.json();
            this.email = data[0];
            [this.login, this.domain] = this.email.split('@');
        } catch (e) {
            throw new Error('Error have occurred while generating email: ' + e.message);
        }
    }

    async doAction(params) {
        const url = new URL('https://www.1secmail.com/api/v1/');
        Object.keys(params).forEach(key => url.searchParams.append(key, params[key]));
        const response = await fetch(url);
        return response.json();
    }

    async fetchEmails(from = null) {
        try {
            const dat = await this.doAction({
                "action": "getMessages",
                "login": this.login,
                "domain": this.domain
            });
            if (from) {
                return dat.filter(ml => ml.from.includes(from))
                    .map(ml => new Email(this.login, this.domain, ml.id, ml.from, this.doAction.bind(this)));
            }
            return dat.map(ml => new Email(this.login, this.domain, ml.id, ml.from, this.doAction.bind(this)));
        } catch (e) {
            throw new Error('Error have occurred while fetching emails: ' + e.message);
        }
    }
}

